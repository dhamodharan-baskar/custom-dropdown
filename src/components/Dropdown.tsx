import { AiOutlineClose, AiOutlineCaretDown } from "react-icons/ai";
import { useEffect, useRef, useState} from 'react'
import styled from "styled-components";
interface DropdownProps {
    isSearchable?: boolean;
    data?: any[];
    inputClass?: object;
    optionOverviewClass?: object;
    dropdownItemClass?: object;
    removeClear?: boolean;
    accessorKey?: string;
    selectedValue?: string;
    onChangeValue?: (value: string) => void;
    onSelectOption?: (value: object) => void;
    autoComplete?: boolean;
    search?:boolean
   }

   export default function Dropdown(props: DropdownProps){
    const [searchValue, setSearchValue] = useState<string>('')
    const [showOptions, setShowOptions] = useState<boolean>(false)
    const [dropdownOptions, setDropdownOptions] = useState<any[]>([])
    const [filterdData, setFilterdData] = useState<any[]>([])
    const dropdownRef = useRef<any>();
    const inputRef = useRef<any>();
    const {
        inputClass,
        optionOverviewClass,
        dropdownItemClass,
        removeClear,
        accessorKey,
        data,
        onChangeValue,
        onSelectOption,
        selectedValue,
        search,
        autoComplete
    } = props;
    let dropdownValue = search ? filterdData : dropdownOptions
    useEffect(() => {
        if(data){
            setDropdownOptions(data);
            setFilterdData(data);
        }
    },[data]);

    useEffect(() => {
        document.addEventListener('mousedown', handleClickOutside);
        return () => document.removeEventListener('mousedown', handleClickOutside);
    });

    useEffect(() => {
        if(selectedValue){
            setSearchValue(selectedValue)
        }
    },[selectedValue]);

    const handleClickOutside = (e: any) => {
        if (!dropdownRef?.current?.contains(e.target) ) {
            setShowOptions(false);
        } 
    };

    const onChangeInput = (value: string) => {
        if(search){
            let key = accessorKey ?? 'value'
            let filterData: any[] = dropdownOptions?.filter((child) => child[key]?.includes(value));
            setFilterdData(filterData)
        }
        setSearchValue(value)
        setShowOptions(true)
        onChangeValue && onChangeValue(value)
    }

    const clearSearch = () => {
        if(search){
            setDropdownOptions(data ?? [])
        }
        setSearchValue('')
        setShowOptions(false)
        onSelectOption && onSelectOption({})
    }

    const onSelect = (value: Object) => {
        onSelectOption && onSelectOption(value);
        setShowOptions(false); 
    }

    const highLightText = (value: string):any => {
      return(
        <span>
            {
             value?.split('').map((item: string, index: number) => {
                return(
                    <span 
                        key={item + index}
                        style={{color: searchValue?.toLowerCase().includes(item?.toLowerCase()) ? 'Highlight' : 'ActiveBorder' }}>
                        {item}
                    </span>
                )
             })
            }
        </span>
      )
    }

    return (
        <DropdownOverview>
            <DropdownInput 
                ref={inputRef}
                data-testid="dropdown-input-overview"
                onClick={() => (!autoComplete) && setShowOptions(!showOptions)}>
                <StyledInput 
                    className={`${inputClass ? inputClass : ''}`}
                    value={searchValue} 
                    data-testid="dropdown-input-box" 
                    disabled = {!search && !autoComplete}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => onChangeInput(event.target.value)}/>
                <Options>
                {!removeClear &&<AiOutlineClose 
                    data-testid="dropdown-clear-options" 
                    onClick={ () => clearSearch()} style={{cursor: 'pointer'}}/>}
                <AiOutlineCaretDown onClick={() => setShowOptions(!showOptions)} style={{cursor: 'pointer'}}/>
                </Options>
            </DropdownInput>
            {
                showOptions &&
                <DropdownOptions 
                    ref={dropdownRef}
                    data-testid="dropdown-options-overview" 
                    className={`${optionOverviewClass ? optionOverviewClass : ''}`}>
                {
                    dropdownValue?.length ?
                    <div>
                        {
                            dropdownValue?.map((item: any, index: number) => {
                                return(
                                    <DropdownItem 
                                        key={`${accessorKey ? item[accessorKey] : item?.value} + ${index}`}
                                        onClick={() => onSelect(item) }
                                        className={`${dropdownItemClass ? dropdownItemClass : ''}`}>
                                        {highLightText( accessorKey ? item[accessorKey] : item?.value)}
                                    </DropdownItem>
                                )
                            })
                        }
                    </div>
                    : 'no results found'
                }
                </DropdownOptions>
            }
        </DropdownOverview>
    )
}

const DropdownOverview = styled.div`
  width: auto;
  min-width: 200px;
`
const DropdownInput = styled.div`
    width: 100%;
    border: 1px solid #D8D8D8;
    border-radius: 4px;
    position: relative;
    padding: 4px 12px;
    display: flex;
    justify-content: space-between;
    margin-bottom: 4px;
    min-height: 32px;
    align-items: center;
    cursor: pointer;
`
const DropdownOptions = styled.div`
  min-width: 200px;
  max-height: 300px;
  overflow: auto;
  border: 1px solid #D8D8D8;
  border-radius: 4px;
  position: absolute;
  background-color: #fff;
  padding: 8px;
  z-index: 999;
`

const DropdownItem = styled.div`
font-size: 16px;
padding: 8px;
cursor: pointer;
border-radius: 4px;
&:hover {
    background: #f2f2f2;
}    
`
const StyledInput = styled.input`
  border: none;
  &:focus {
    outline: none;
}
`
const Options = styled.div`
display: flex;
justify-content: space-between;
gap: 4px;
`