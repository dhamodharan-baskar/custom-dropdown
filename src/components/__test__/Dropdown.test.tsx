import { fireEvent, render } from "@testing-library/react";
import { Fragment } from "react";
import Dropdown from '../Dropdown'
const dropdownData = [
    {id: 1, value: 'user1', label: 'user1-label'}, 
    {id: 2, value: 'user2', label: 'user1-label'}, 
    {id: 3, value: 'user3', label: 'user1-label'},
    {id: 4, value: 'user4', label: 'user1-label'},
    {id: 5, value: 'user5', label: 'user1-label'}
  ]

describe("dropdown", () => {
    it("renders the dropdown component", () => {
      render(<Dropdown />);
      expect(Fragment).toMatchSnapshot();
    });
    it("renders the dropdown input", () => {
        const {getByTestId} = render(<Dropdown />);
        const element = getByTestId('dropdown-input-box')
        expect(element).toBeInTheDocument;
      });
      it("renders the clear option", () => {
        const {getByTestId} = render(<Dropdown />);
        const element = getByTestId('dropdown-clear-options')
        expect(element).toBeInTheDocument;
      });
      it("renders the options", () => {
        const {getByTestId} = render(<Dropdown data={dropdownData}/>);
        const element = getByTestId('dropdown-input-overview')
        fireEvent.click(element);
        expect(getByTestId('dropdown-options-overview')).toBeInTheDocument;
      });
      it("renders the input", () => {
        const {getByTestId} = render(<Dropdown data={dropdownData}/>);
        const element = getByTestId("dropdown-input-box" )
        expect(element).toBeInTheDocument;
      });
      it("renders the input to be disabled", () => {
        const {getByTestId} = render(<Dropdown data={dropdownData} />);
        const element = getByTestId("dropdown-input-box" )
        expect(element).toBeDisabled;
      });
      it("renders the input to be enabled", () => {
        const {getByTestId} = render(<Dropdown data={dropdownData} search/>);
        const element = getByTestId("dropdown-input-box" )
        expect(element).toBeEnabled;
      });
  });

  