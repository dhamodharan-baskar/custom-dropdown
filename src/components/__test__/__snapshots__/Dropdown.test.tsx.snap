// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown renders the correct number of stars 1`] = `Symbol(react.fragment)`;

exports[`dropdown renders the dropdown component 1`] = `Symbol(react.fragment)`;
